docs/src/errortypes.md

```diff
@@ -0,0 +1,263 @@
+end
+
+julia> # Errors if x.x isa String.
+       # The compiler doesn't know if it's a String or Int
+       f(x) = x.x + 1;
+
+julia> @report_call f(Foo(1))
+═════ 1 possible error found ═════
+┌ @ REPL[2]:1 x.x + 1
+│ no matching method found `+(::String, ::Int64)` (1/2 union split): ((x::Foo).x::Union{Int64, String} + 1)
+└─────────────
+```
+
+More common example:
+```julia
+julia> function pos_after_tab(v::AbstractArray{UInt8})
+    # findfirst can return `nothing` on no match
+    p = findfirst(isequal(UInt8('\t')), v)
+    p + 1
+end
+
+julia> @report_call pos_after_tab(codeunits("a\tb"))
+═════ 1 possible error found ═════
+┌ @ REPL[17]:2 p + 1
+│ no matching method found `+(::Nothing, ::Int64)` (1/2 union split): (findfirst(isequal(UInt8)::UInt8)::Base.Fix2{typeof(isequal), UInt8}, v::Base.CodeUnits{UInt8, String})::Union{Nothing, Int64} + 1)
+└──────────────
+```
+
+### How to fix
+This error is unique in that idiomatic Julia code may still lead to this error. For example, in the `pos_after_tab` function above, if the input vector does not have a `'\t'` byte, `p` will be `nothing`, and a `MethodError` will be thrown when `nothing + 1` is attempted. However, in many situations, the possibility of such a `MethodError` is not a mistake, but rather an idiomatic way of erroring.
+
+There are different possiblities to address this kind of error. Let's take the `pos_after_tab` example:
+
+If you actually _could_ expect `p` to legitimately be `nothing` for valid input (i.e. there could be no `'\t'` byte), then your function should be written to take this edge case into account:
+```julia
+julia> function pos_after_tab(v::AbstractArray{UInt8})
+    p = findfirst(isequal(UInt8('\t')), v)
+    if p === nothing # handle the nothing case
+        return nothing
+    else
+        return p + 1
+    end
+end;
+
+julia> @report_call pos_after_tab(codeunits("a\tb"))
+No errors detected
+```
+
+By adding the `if p === nothing` check, the compiler will know that the type of `p` must be `Nothing` inside the `if` block, and `Int` in the else block. This way, the compiler knows a `MethodError` is not possible, and the error will disappear.
+
+If you expect a `'\t'` byte to always be present, such that `findfirst` always should return an `Int` for valid input, you can add a typeassert in the function to assert that the return value of `findfirst` must be, say, an `Integer`. Then, the compiler will know that if the typeassert passes, the value returned by `findfirst` cannot be `nothing` (and hence in this case must be `Int`):
+
+```julia
+julia> function pos_after_tab(v::AbstractArray{UInt8})
+    p = findfirst(isequal(UInt8('\t')), v)::Integer
+    p + 1
+end
+
+julia> @report_call pos_after_tab(codeunits("a\tb"))
+No errors detected
+```
+
+The code will still error at runtime due to the typeassert if `findfirst` returns `nothing`, but JET will no longer detect it as an error, as violating a typeassert is not considered a type error to JET.
```

Comment:
> violating a typeassert is not considered a type error to JET

That is concerning.

Reply:
Not really, though I get why you would think that. If I do the following:
```julia
p = findfirst(pred, xs)::Int
```
Then Julia will throw if `findfirst` does not return an `Int`. This throw is not really an issue we would want a tool like JET to complain about, because the typeassert means that the programmer explicitly said "I know this might not be inferred precisely, but I _guarantee_ it's an Int, so the compiler can assume that `p` is an `Int`, no matter what type inference says". Having JET complain about it would not help anything, since the programmer already know it might not be inferred to `Int`, else they would not add the typeassert.

Note that the code is still safe in the sense that if the programmer is wrong, and it does not return an `Int`, the function will throw an exception. It will not lead to miscompilations.